Changelog.md

```diff
@@ -10,6 +10,7 @@ Compiler Features:
 
 
 Bugfixes:
+ * Name Resolver: Prevent member/parameter names from shadowing type names within struct/enum/error/event declarations.
```

Comment:
Do we issue an error in case of shadowing? I'm trying to understand how we prevent shadowing, because users could still do it, right?

Reply:
There's no error. That's the point of the PR. There are some examples in the original issue (#9231) that should make this clearer.

For example before this PR you would get an error here:
```solidity
enum E {a, b, c}
struct T {E T; uint E;}
```
```
Error: Name has to refer to a struct, enum or contract.
 --> test.sol:2:11:
  |
2 | struct T {E T; uint E;}
  |           ^
```

That's because the name checker thinks that `uint E` is shadowing the type `E` inside the body of the struct.

The PR changes that by making name checker aware that names defined in some kinds of scopes (struct declarations, enum declarations, function declarations without a body, event/error declarations) cannot shadow anything **inside that declaration**.

Reply:
@cameel 
> Prevent member/parameter names from shadowing type names within struct/enum/error/event declarations.

I think, strictly speaking, we do not _prevent shadowing_, we rather _prevent recognition of non-existing shadowing as shadowing_. Maybe

> Allow member/parameter names to be the same as type names within struct/enum/error/event declarations (avoid false name shadowing errors)

BTW, thanks a lot for taking care of the PR! (I still do hope to come back to the project for real eventually, even if right now it does not sound realistic.)

Reply:
Well, yeah, I get your point but I have a really hard time expressing this in a changelog entry without making it confusing to users or taking a whole paragraph :) This is the best I could come up with. If anyone has a better suggestion, I can replace it.

> BTW, thanks a lot for taking care of the PR! (I still do hope to come back to the project for real eventually, even if right now it does not sound realistic.)

Nice to heard that. I hope we'll get you back at some point :)